storage/simulation/script.go

```diff
@@ -0,0 +1,310 @@
+)
+
+var operations = [...]string{
+	OpSplitRange: "splitrange",
+	OpAddNode:    "addnode",
+	OpExit:       "exit",
+}
+
+// String returns the string version of the Operation enumerable.
+func (operation Operation) String() string {
+	return operations[operation]
+}
+
+// operationFromString returns the operation corresponding with the passed in
+// string.
+func operationFromString(s string) (Operation, bool) {
+	for i, op := range operations {
+		if op == s {
+			return Operation(i), true
+		}
+	}
+	return 0, false
+}
+
+// OperationValue enumerates the possible values for a scripted operation.
+type OperationValue int
+
+// These are the possible operations values.
+const (
+	OpValSet OperationValue = iota
+	OpValRandom
+	OpValFirst
+	OpValLast
+)
+
+var operationValues = [...]string{
+	OpValSet:    "",
+	OpValRandom: "random",
+	OpValFirst:  "first",
+	OpValLast:   "last",
+}
+
+// String returns the string version of the OperationValue enumerable.
+func (operationValue OperationValue) String() string {
+	return operationValues[operationValue]
+}
+
+// operationValueFromString returns the operation corresponding with the passed
+// in string. If the value is opValSet, then the value is parsed into the
+// returned int.
+func operationValueFromString(s string) (OperationValue, int, error) {
+	for i, opVal := range operationValues {
+		if opVal == s {
+			return OperationValue(i), 0, nil
+		}
+	}
+	valueNumber, err := strconv.Atoi(s)
+	return OpValSet, int(valueNumber), err
+}
+
+// Action is a single scripted action.
+type Action struct {
+	operation   Operation
+	value       OperationValue
+	valueNumber int
+}
+
+// ActionDetails contains an action and all of the metadata surrounding that
+// action.
+// First is the first epoch in which the action occurs.
+// Every is the interval for how often the action re-occurs.
+// Last is the final epoch in which the action can occur.
+// Repeat is how many times an action occurs at each epoch in which it fires.
+type ActionDetails struct {
+	Action
+	first  int
+	every  int
+	last   int
+	repeat int
+}
+
+func (ad ActionDetails) String() string {
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "Operation:%s", operations[ad.Action.operation])
+
+	if ad.Action.value == OpValSet {
+		if ad.Action.valueNumber > 0 {
+			fmt.Fprintf(&buf, "\tValue:%d", ad.Action.valueNumber)
+		} else {
+			fmt.Fprintf(&buf, "\t ")
+		}
+	} else {
+		fmt.Fprintf(&buf, "\tValue:%s", operations[ad.Action.operation])
+	}
+
+	fmt.Fprintf(&buf, "\tFirst:%d", ad.first)
+	if ad.last > 0 {
+		fmt.Fprintf(&buf, "\tLast:%d", ad.last)
+	} else {
+		fmt.Fprintf(&buf, "\t")
+	}
+	if ad.every > 0 {
+		fmt.Fprintf(&buf, "\tEvery:%d", ad.every)
+	} else {
+		fmt.Fprintf(&buf, "\t")
+	}
+	if ad.repeat > 1 {
+		fmt.Fprintf(&buf, "\tRepeat:%d", ad.repeat)
+	}
+	return buf.String()
+}
+
+// Script contains a list of all the scripted actions for cluster simulation.
+type Script struct {
+	maxEpoch int
+	actions  map[int][]Action
```

Comment:
this representation implies that multiple actions are allowed to take place in each epoch. is this intentional? ISTM that actions are not commutative. In particular, this representation allows for violating the invariant that each range is allowed to act at most once per epoch.